pkg/chunk/cache/redis_client.go

```diff
@@ -0,0 +1,304 @@
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	cmd := c.rdb.MGet(ctx, keys...)
+	if err := cmd.Err(); err != nil {
+		return nil, err
+	}
+
+	ret := make([][]byte, len(keys))
+	for i, val := range cmd.Val() {
+		if val != nil {
+			ret[i] = StringToBytes(val.(string))
+		}
+	}
+	return ret, nil
+}
+
+func (c *redisBasicClient) Close() error {
+	return c.rdb.Close()
+}
+
+type redisSentinelClient struct {
+	expiration time.Duration
+	timeout    time.Duration
+	rdb        *redis.SentinelClient
+	opt        *redis.Options
+}
+
+func newRedisSentinelClient(cfg *RedisConfig) *redisSentinelClient {
+	opt := &redis.Options{
+		Addr:        cfg.Endpoint,
+		Password:    cfg.Password.Value,
+		PoolSize:    cfg.PoolSize,
+		IdleTimeout: cfg.IdleTimeout,
+		MaxConnAge:  cfg.MaxConnAge,
+	}
+	if cfg.EnableTLS {
+		opt.TLSConfig = &tls.Config{}
+	}
+	return &redisSentinelClient{
+		expiration: cfg.Expiration,
+		timeout:    cfg.Timeout,
+		rdb:        redis.NewSentinelClient(opt),
+		opt:        opt,
+	}
+}
+
+func (c *redisSentinelClient) getMaster(ctx context.Context) (*redisBasicClient, error) {
+	masters, err := c.rdb.Masters(ctx).Result()
+	if err != nil {
+		return nil, err
+	}
+	if len(masters) == 0 {
+		return nil, ErrNoMasters
+	}
+	for _, master := range masters {
+		// expected: master = []interface {}{"name", "<master name>", "ip", "<IP>", "port", "<port>", ... }
+		if info, ok := master.([]interface{}); ok {
+			if len(info) >= 6 && info[2] == "ip" && info[4] == "port" {
+				return &redisBasicClient{
+					expiration: c.expiration,
+					timeout:    c.timeout,
+					rdb: redis.NewClient(&redis.Options{
+						Addr:        fmt.Sprintf("%s:%s", info[3], info[5]),
+						Password:    c.opt.Password,
+						PoolSize:    c.opt.PoolSize,
+						IdleTimeout: c.opt.IdleTimeout,
+						MaxConnAge:  c.opt.MaxConnAge,
+					}),
+				}, nil
+			}
+		}
+	}
+	return nil, ErrNoMasters
+}
+
+func (c *redisSentinelClient) Ping(ctx context.Context) error {
+	var cancel context.CancelFunc
+	if c.timeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, c.timeout)
+		defer cancel()
+	}
+
+	pong, err := c.rdb.Ping(ctx).Result()
+	if err != nil {
+		return err
+	}
+	if pong != "PONG" {
+		return fmt.Errorf("redis: Unexpected PING response %q", pong)
+	}
+	return nil
+}
+
+func (c *redisSentinelClient) MSet(ctx context.Context, keys []string, values [][]byte) error {
+	master, err := c.getMaster(ctx)
+	if err != nil {
+		return err
+	}
+	return master.MSet(ctx, keys, values)
+}
+
+func (c *redisSentinelClient) MGet(ctx context.Context, keys []string) ([][]byte, error) {
+	master, err := c.getMaster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return master.MGet(ctx, keys)
+}
+
+func (c *redisSentinelClient) Close() error {
+	return c.rdb.Close()
+}
+
+// StringToBytes converts string to byte slice. (copied from vendor/github.com/go-redis/redis/v8/internal/util/unsafe.go)
```

Comment:
Could we just use the one from `vendor/github.com/go-redis/redis/v8/internal/util/unsafe.go`?

Reply:
Go throws an error when you import internal package: "use of internal package github.com/go-redis/redis/v8/internal/util not allowed"